### DISCUSSION TITLE
[RAWeb API] Way to associate RAWeb `gameId` to Local games?

### DISCUSSION THREAD

**Discussion by TapiocaFox** [RetroAchievements/RAWeb#2617] [Ideas]
Hello, I am TapiocaFox. Thanks for your amazing works revive retro gaming!

In Daijishou there was a feature where user can import games. The launcher then used "the title of the game" to query `retroachievements.org` and find correct `gameId` to associate game to RA. Which displayed the relevant RA information for user to access with ease.

Since the site is protected by anti-bot and no longer scrapable. I am revamping achievements features using RAWeb API. However, I didn't see similar alternatives in the API. Maybe I missed something? If not how do I provide the same feature possibly? (Hash is kinda slow?)

Thanks for your time!

![image](https://github.com/user-attachments/assets/ab030082-4e22-4846-8f0f-bcfb03785381)

![image](https://github.com/user-attachments/assets/501ea5ee-2588-470a-98f1-bc46410ef7f6)

**Comment by wescopeland** [comment]
Hello again, @TapiocaFox!

At the moment, https://api-docs.retroachievements.org/v1/get-game-list.html is probably the closest thing I can think of that does something like what you're asking for, but I suspect this doesn't satisfy your needs.

I'd love to help you on this, if what you're asking for is something we can do. Especially if it means restoring RetroAchievements functionality in Daijishou.

Are you specifically looking for an endpoint that returns a game ID if you provide a hash?

**Reply by TapiocaFox** [reply]
Hey, nice to see you again! Thanks for the orientation. I am aware of `get-game-list` endpoint! 

However, the mechanism Dai used to link local games to RA's `gameId` was based on "**game title**" purely. Namely "https://retroachievements.org/searchresults.php?s=". Therefore what I need is more like a search by "game title" feature filtered by platform type. 

There are reasons for this. Future proof, extendable, speed, complexity.
1. "**Game title**" method is more extendable and future proof to fetch many information (other than retroachievemts) for lots of rare, new or custom platforms that is not as well indexed by online community. Means that even if Daijishou app isn't updated. The community can just modify the config file to get what they need. The mechanism Dai provided is platform-independent (not determined by which gaming).
2. Platforms are defined by config files in Daijishou and are not hardcoded. If each platform has its own hashing methods. It is  *coupled with Daijishou internal "hashing availability", *tedious to implement and *hard to generalize.
3. I've found hashing is rather demanding for synchronization, users may have thousands of items and large files on slower devices. In Daijishou, `LazyMd5` only hash first 4MB of the file to guarantee consistent sync speed. 
4. The hash use cases are rare in Daijishou. Only used to resync file or prevent duplicated files synced differently in to the library.

Regardless, Maybe I should rethink more. But these are my current obstacles.

**Reply by zikiri33** [reply]
I'm confused. Your method has already proven to not be future proof, isn't that why you're here? And without RA's hashing, you would just be connecting files to Retro Achievement sets and telling users the sets and files match, even if they don't. Is that right?

But if you do want to do this, the getGameList endpoint provides every game title and their matching hashes. And you could use the data from this repo to even check nointro/redump filenames https://github.com/RetroAchievements/RAHashes 

(Also for the record I'm not associated with the RA team at all. I'm just really confused about how you intend to futureproof something that depends on frequently updated data from 3rd parties)

**Reply by TapiocaFox** [reply]
Your argument not really make sense @zikiri33. First, everything is working and maintained by the community except for RetroAchievements to now. The thing really hold Daijishou's RA system back is anti-bot implementation on the RA website which does not involve with the discussion. I am currently proposing "search by title" method I under my "full commitment to RAWeb API".

In the past, Dai is the first Android frontend that utilize RA, and at the time there is no API for kotlin etc. Busy along with other tasks of the project, I decided to use the scraping method, since there are lots of features not supported by API 2 years ago yet (even now). For example, friend list and other social features. In the end, the method matched around 95% correctly. Very small portion of users had complain at the time. It was well established and people using it regularly.

Now I am switching to use RAWeb API. I am just looking for the possibilities that are compatible with my already established works with my thoughts. And seek convenient changes. Not saying my past implementation is right, or wrong.

**Reply by wescopeland** [reply]
Hi @TapiocaFox!

I've been mulling over how we can help you here :-)

Unfortunately, a search endpoint is one thing I think for the time being we won't be able to expose to the web API 😔 The reason why is our on-site search functionality is a bit naive; it relies entirely on SQL `LIKE` string comparisons, which unfortunately bypass indexing on our tables. In other words: it's quite slow and computationally-intensive.

Given that constraint, I'm wondering if there's any other way we can find a happy medium here that can help Daijishou's integration. Certainly open to ideas 🙂

**Reply by TapiocaFox** [reply]
@wescopeland Got it, thanks for your time! I will see If I can work it out. Do you suggest any library in any language that implements RA's hashes as a paradigm. So maybe I can implement RA hashes myself with a reference?

**Reply by wescopeland** [reply]
The only thing I know of is https://github.com/RetroAchievements/rcheevos itself -- this low-level library is what emulators use to hash ROMs.

I might also recommend perhaps investigating [Batocera's RetroAchievements integration](https://github.com/batocera-linux/batocera-emulationstation/blob/3959d0742f1b3fb9e3b86e56f365d5eea99eaf5d/es-app/src/RetroAchievements.cpp#L595C11-L595C11), which I believe uses rcheevos to hash the user's ROM library. Unfortunately this code is a little outside of my wheelhouse so I can't offer any direct support 😅, but it may help inspire some ideas on how to tackle the problem as I believe Batocera shares a very similar problem space with what you're trying to solve.

**Reply by TapiocaFox** [reply]
Thanks a lots. I think I've seen https://github.com/RetroAchievements/rcheevos before. I will dive in more. Hope my implementation not caught on fire. 😂

**Reply by TapiocaFox** [reply]
I found some hashes require domain knowledge of emulation. Quite a steep learning curve and might even need to implement a simple emulator. For example:

### GameCube uses a custom hash. Initially loaded program code is hashed as follows:
1. The Apploader is loaded and added to a buffer.
2. Each DOL segment (code and data) referenced in the Apploader is loaded and added to the buffer.
3. The buffer is hashed.

Looking into C++ code there are 4k line of code for hashing, doubt I will ever be able to finish it all 😨:

https://github.com/RetroAchievements/rcheevos/blob/f1d6965e9bc7be497182d9d66087086df9137b49/src/rhash/hash.c#L4097

I wonder whether RA devs can help with Java/Kotlin Library for hashing.
